### DISCUSSION TITLE
dj64dev introduction video

### DISCUSSION THREAD

**Discussion by stsp** [dosemu2/dosemu2#2333] [General]
https://www.patreon.com/posts/introduction-to-118824299?utm_medium=clipboard_copy&utm_source=copyLink&utm_campaign=postshare_creator&utm_content=join_link

I made an introduction video for dj64dev.
Its not very useful, just shows how to build
ncurses. This is the first time ever I spoke
English, first time ever I made any video,
and obviously I had some problems with
microphone, as the sound quality is very
low. Not sure why is so.
In fact, initially I couldn't even run OBS:
https://bugzilla.redhat.com/show_bug.cgi?id=2334607
But after installing it from flatpak, it worked.
Obviously when you do something for
the first time, you hit all the possible
problems at once. :)

**Comment by jschwartzenberg** [comment]
> This is the first time ever I spoke
> English, first time ever I made any video,
> and obviously I had some problems with
> microphone, as the sound quality is very
> low. Not sure why is so.

Congrats on this first step! Your pronunciation is quite good, I've heard much stronger accents!!
You can probably increase the sound quality of the recording using Audacious. Record a bit of the noise only and then apply noise reduction to the entire track using that bit as input. I've managed to remove noise from badly recorded audio in the past myself with that as well.

Looking forward to see more videos! Maybe we could share them on more sites to increase the visibility of the project. Do you have any Fediverse account?

**Reply by stsp** [reply]
Thanks!
Its not like I am going to be a
youtuber, hopefully. :) So perhaps
there would be no more videos. I
just mentioned on Patreon page
that I am writing this tool-chain,
so the video was a due to those
who supported.

> You can probably increase the sound quality of the recording using Audacious.

But I am not recording the audio separately,
so will audacious work with video track?
Anyway, I can clearly hear echo, and I
think the most stupid thing was that I
didn't switch off the loudspeakers in front
of myself. They likely created echo...
The noise seems to be specific to onboard
audio: i tried USB mic and it seems much
better. So there might be a problem with
drivers perhaps, not sure.

> Looking forward to see more videos!

On what topics?

> Do you have any Fediverse account?

No, I just today created the vimeo account
for that video, and I have nothing else.

**Reply by stsp** [reply]
> On what topics?

Its actually not a rhetorical question.
I have big difficulties coming up with
an interesting events related to dosemu2.
If/when it works under browser, it will
probably be worth a video (but dosbox
can already do that I suppose).
Or we could, for example, turn it into
a TAS system (for tool-assisted speedruns).
That would definitely produce a lot of
fun and videos, but its not a simple task.

And I don't know what else could be
interesting. Ideas welcome. :)

**Reply by ecm-pushbx** [reply]
I'd like to have a rewind / replay feature that allows debugging backwards in time. This is called time travel, and it can be done with a VM but may require a lot of memory. Also, the debugger interface that controls it probably needs to be outside of the VM. (I do much prefer using lDebug in a dosemu2 rather than using dosemu2's dosdebug.)

**Reply by stsp** [reply]
> I'd like to have a rewind / replay feature that allows debugging backwards in time.

Yep, that's an essential part of the
TAS framework, so we are talking
about the same thing. Unfortunately
its exceptionally difficult to implement,
especially if you want a fine-granularity,
eg move backwards 2 or 3 frames, rather
than the whole 10 seconds to prev snapshot.

> This is called time travel, and it can be done with a VM but may require a lot of memory.

... and most likely can only be done on interpreter.

> Also, the debugger interface that controls it probably needs to be outside of the VM.

What exactly do you mean?
dosdebug is a standalone tool that
works outside of a VM, and communicates
to it with pipes.

**Reply by ecm-pushbx** [reply]
> What exactly do you mean?
dosdebug is a standalone tool that
works outside of a VM, and communicates
to it with pipes.

Yes, I meant that would work for time travel. lDebug wouldn't because as a DOS application it also runs in the same environment as the debuggee. I wish I could use lDebug though because its run commands (T/TP/P/G) and breakpoints are better than dosdebug's.

**Reply by stsp** [reply]
In what way better?
Although I know that dosdebug
is quite buggy, as no one invests
so many efforts in it as you do in
IDebug.

**Reply by jschwartzenberg** [reply]
> But I am not recording the audio separately,
> so will audacious work with video track?

No you'll have to separate the audio track and then add the fixed up version to the video track again. This should not be too difficult, especially as the length will be the same. There are tools to do this, I think `ffmpeg` for instance can do this.


> The noise seems to be specific to onboard
> audio: i tried USB mic and it seems much
> better. So there might be a problem with
> drivers perhaps, not sure.

Maybe also check if there are multiple capture inputs enabled when recording. Make sure all but the microphone are disabled. You also want to disable the playback input of the microphone. Indeed USB headsets/microphones tend to be much better nowadays.


> On what topics?

For me personally anything that you fix would be nice to see with a before and after video. Especially when it's a nice game ;) Videos do not have to be as long as the one you made now. It could also just be a few minutes demonstrating just a particular improvement. Maybe in such videos you could also explain how you managed to find and solve an issue in the code so that people can learn from your debugging strategies.


> > Do you have any Fediverse account?
> 
> No, I just today created the vimeo account for that video, and I have nothing else.

Maybe you could look into it. Most sites within the Fediverse are running Mastodon. You can find some info here: https://joinmastodon.org/
Somebody also created Smithereen which looks like VK: https://friends.grishka.me/
As all hosts support ActivityPub, you can follow people on other servers.

Some other accounts which focus on DOS stuff:
https://bitbang.social/@loadhigh
https://mastodon.social/@dosnostalgic

**Reply by stsp** [reply]
> Maybe also check if there are multiple capture inputs enabled when recording.

On that device, the rest is "unplugged"
in pavucontrol. But indeed I need to try
muting some things to make sure.

> For me personally anything that you fix would be nice to see with a before and after video. Especially when it's a nice game ;) 

These are totally different animals. :)
Most of things I fix, are some configure
problems or dj64dev problems. Only
perhaps 1% of fixes is targeted to "some
game".

> Maybe in such videos you could also explain how you managed to find and solve an issue in the code so
> that people can learn from your debugging strategies.

I don't think this kind of audience is
large enough. Of course @ecm-pushbx 
will look how am I debugging something,
as she's an author of a debugger. No one
else will. :)
There are many technically cool things
that no one will want to understand.

> As all hosts support ActivityPub, you can follow people on other servers.

What's the purpose?
I registered on vimeo because this is
what patreon uses by default. What
other hostings can give above that?

> Some other accounts which focus on DOS stuff:

Nice places. :)
Will keep an eye on them.

**Reply by jschwartzenberg** [reply]
> What's the purpose?

Essentially you answered that below:

> Will keep an eye on them.

You can keep an eye on them and interact with them when you have an account :)


> These are totally different animals. :)
> Most of things I fix, are some configure
> problems or dj64dev problems. Only
> perhaps 1% of fixes is targeted to "some
> game".

That's true, but I think a series of short videos on DOSEMU2 development would attract new audiences. There are also quite some videos on YouTube of people showing off their debugging skills. But having such videos for DOSEMU2 would lower the barrier for others to join in development.


> Of course @ecm-pushbx
> will look how am I debugging something,
> as she's an author of a debugger. No one
> else will. :)

I will ;) And I'm also ready to watch @ecm-pushbx's videos :) :)

**Reply by ecm-pushbx** [reply]
> In what way better? Although I know that dosdebug is quite buggy, as no one invests so many efforts in it as you do in IDebug.

The segmented address input seems to be pickier than what I'm used to if I recall. Is there a P command (proceed past call or loop instructions) in dosdebug at all? Running a small test I cannot manage to get the BP + G commands to work as I want reliably. G command seems to run past the breakpoints without hitting them a lot.

**Reply by ecm-pushbx** [reply]
> > Of course @ecm-pushbx
> > will look how am I debugging something,
> > as she's an author of a debugger. No one
> > else will. :)
> 
> I will ;) And I'm also ready to watch @ecm-pushbx's videos :) :)

Oh no, I much prefer the written word. Haven't watched @stsp's video yet either. If you have suggestions for things I could cover maybe I'll slot them into blog posts. Do you know about my blog yet? It's at https://pushbx.org/ecm/dokuwiki/

**Reply by stsp** [reply]
> G command seems to run past the breakpoints without hitting them a lot.

This is likely a bug that should be
reported. For the rest - it shouldn't
be difficult to add, but currently there
is probably no one interested enough
in such functionality, to spend the
efforts.

**Reply by stsp** [reply]
> Oh no, I much prefer the written word.

Some things are better be shown.
Eg if you added this time-travel thing,
what's the use of writing? Just show
the TAS with it. :)
But you are right that most things are
better be written instead.

**Reply by stsp** [reply]
> some videos on YouTube of people showing off their debugging skills.
> But having such videos for DOSEMU2 would lower the barrier for others to join in development.

That's a bit of a wrong understanding.
dosemu2 has no high barrier where you
need to be able to rev-engineer the DOS
games before contributing, no! As I said,
only maybe 1% of my work is targeted
towards debugging some game. In dosemu2
there are the infinite amount of areas to
work on, each requiring different skill levels.
So... the thing is: there are just no contributors. :)
And the technical barrier is not a problem here.

**Reply by ecm-pushbx** [reply]
> In what way better? Although I know that dosdebug is quite buggy, as no one invests so many efforts in it as you do in IDebug.

It looks almost the same on the github page but my lDOS programs are "small L" DOS, not "capital i" DOS ;)

**Reply by stsp** [reply]
> Do you know about my blog yet? It's at https://pushbx.org/ecm/dokuwiki/

I believe the question wasn't to me,
but I wasn't aware of your blog. I looked
now, and it seems it is mostly related to
your tools, so my guess is that it is only
interesting to your users. This is something
I don't want to do for dosemu2, as dosemu2
doesn't have many users, and not all of the
existing ones are interested in what was changed.
They are mainly interested to not get a new
regression, at best. :)
Finding a cool and interesting topics for
wider audience, is an art that lays too far
from programming. Maybe debugging
games and then playing them a bit, can
indeed reach some audience. But in most
cases you first need to identify the cool
feature, then code it up (like eg TAS system),
then demonstrate. And the second step
(code it up) makes it all non-viable. :)

**Reply by jschwartzenberg** [reply]
> Do you know about my blog yet? It's at https://pushbx.org/ecm/dokuwiki/

Thanks for sharing! I'll see if I can understand some of it. It seems some of your work is captured over multiple posts, but one needs to read them all to have a full understanding. I don't know if there could be a better strategy on organizing all the information. It seems that it can be very helpful for others who want to port things from MASM to NASM.

I think having updates about your blog be available through the Fediverse could help to reach more people with the same interest.


> So... the thing is: there are just no contributors. :)
> And the technical barrier is not a problem here.

I think this is subjective :) I do think your videos could inspire people!


> Maybe debugging
> games and then playing them a bit, can
> indeed reach some audience.

Yes, especially if you manage to put in it a shorter video. Shorter videos tend to be quite popular nowadays.

**Reply by stsp** [reply]
> I think this is subjective :) I do think your videos could inspire people!

But you can do that too, by demonstrating
whatever dosemu2 already has. And we'll
see how it goes. :)

**Reply by jschwartzenberg** [reply]
> Nice places. :)
> Will keep an eye on them.

Another interesting account right now, especially now with a lot of posts related to Windows 2.0 and also driver development for it, see: https://tech.lgbt/@nina_kali_nina/113708472525012650

https://github.com/ninakali/win2stub_dd

**Reply by stsp** [reply]
> https://tech.lgbt/

tech.lgbt? :)
Quite cool blog, non-DOS-related seemingly.
I wonder how many readers do these blogs
have.

**Reply by jschwartzenberg** [reply]
There is a number of followers at the top, but when you follow many people their posts are combined in a single feed and you might not end up reading all of them. I really suggest you also create an account!

My account is here but it is quite mixed regarding topics: https://mas.to/@jschwart

**Reply by jschwartzenberg** [reply]
> tech.lgbt? :)

Note that that is just one Mastodon instance. One of many. Anybody can create an account on any and follow people on any other server as I mentioned earlier all Mastodon instances are part of the Fediverse.

**Reply by stsp** [reply]
Are you really working on ASML company?
Though some googling reveals, it also may
be Abstract State Machine Language. :)

**Reply by jschwartzenberg** [reply]
> Are you really working on ASML company?

I'm at this one: https://www.asml.com
There's also a bit of public info on Luxoft's relation with ASML here: https://career.luxoft.com/locations/netherlands

**Reply by stsp** [reply]
Cool!
I hope your contributions to dosemu2
will then not be limited to python scripting
only. :)

**Reply by jschwartzenberg** [reply]
I actually just joined a new team that's mainly using Python ;)

But yeah my interest is actually more low level. I've been looking at more C code lately and also some Udemy courses. If there are any topics you can recommend to explicitly study, I'd be curious. Maybe x86 assembly?

**Reply by stsp** [reply]
> Maybe x86 assembly?

Nope. :)
The last thing the one would study
in 21st century, is x86 asm, so just
don't waste your time on that.
Even to contribute to dosemu2 you
don't need that, as the x86 asm is
probably 0.1% of the entire code base.

And I don't know of any particular topic,
either, because the tasks within dosemu2
range from portability (which means
autoconf, make and scripting) to
parallel programming or JIT engine.
So I think it doesn't matter what skill you
have: if you have a good skill in some
low-level programming area, you can
contribute.

**Reply by stsp** [reply]
Of course the most demanding areas
in dosemu2 are vgaemu and jit - these
were maintained by Bart, but I don't
know where is he now.
These areas indeed require some quite
specific skills to get them working very
fast. Currently they are working as fast
as my skills&time permitted, which means
not fast at all. But I don't know how many
topics you need to study before you can
seriously contribute to those. :)

**Reply by jschwartzenberg** [reply]
Maybe I'll look into graphics programming at some point then....

Btw. another interesting Fediverse profile: https://wuffs.org/@Ninji

He is trying to fix @Baron-Von-Riedesel's patched Windows 3.1 video driver on an eeePC: https://wuffs.org/blog/windows-3x-graphics

**Reply by stsp** [reply]
Yes, but ended up using this one instead:
https://github.com/PluMGMK/vbesvga.drv

**Comment by stuaxo** [comment]
> https://www.patreon.com/posts/introduction-to-118824299?utm_medium=clipboard_copy&utm_source=copyLink&utm_campaign=postshare_creator&utm_content=join_link
> 
> I made an introduction video for dj64dev. Its not very useful, just shows how to build ncurses. This is the first time ever I spoke English, first time ever I made any video, and obviously I had some problems with microphone, as the sound quality is very low. Not sure why is so. In fact, initially I couldn't even run OBS: https://bugzilla.redhat.com/show_bug.cgi?id=2334607 But after installing it from flatpak, it worked. Obviously when you do something for the first time, you hit all the possible problems at once. :)

Impressive, yes - it's amazing how easy some of the older channels make it look, I'm going to make some time to watch this having sampled the first few minutes.

**Reply by stsp** [reply]
> Impressive, yes - it's amazing how easy some of the older channels make it look

What older channels?

**Reply by stsp** [reply]
You mean, how easy it is for people
who do that regularly? Ok, indeed. :)

**Comment by stsp** [comment]
> That's true, but I think a series of short videos on DOSEMU2 development would attract new audiences.
> There are also quite some videos on YouTube of people showing off their debugging skills.

Now as some analytics is available,
I can say that no one have watched
past the demo. Everyone stopped
at "this is all for the demo, next goes
the technical part, and for those who
don't need technical part - thanks for
watching".
Well, I can't claim that for sure, but
since the "Avg time watched" is
exactly 11.10, its pretty much obvious.

So attracting wider audience with
technical details is just not realistic -
in fact, even our small audience of
tech people can't be attracted by that. :)
I think this is just a fact, and neither
me nor @ecm-pushbx nor someone
else who can produce blog or a video,
can change that.

**Reply by andrewbird** [reply]
Oh to be an outlier, I watched to within 3 mins of the end, at which point I was interrupted, but I intend to finish it later. :smile:. I wondered if it's suitable to be used to create a gnumake, and then how the performance would compare to the usual dj dpmi make.

**Reply by stsp** [reply]
Ah, thanks for watching, and thanks
for that info! In fact, vimeo shows 6
watches and patreon shows 9, which
means, somehow vimeo doesn't account
patreon views. And I don't think patreon
offers any analytics... So I don't know
whether people skipped the technical
part following my patreon link. :(
But this also means that 6 people
somehow found that video on their
own, and watched the demo part fully,
without any involvement into dosemu2
community and without any idea what
that video is all about! These are the
true heroes. :)

**Comment by stsp** [comment]
I think the performance of gnu make
would be slightly better with dj64dev:
more memory, more registers, more
optimizations. But given that you
still can't use -j I dont think the
improvement is going to be big.
Note that things are much worse
with fdpp: 64bit mode with flat
pointers is alien to it, so it takes
penalties for frequent switching to
RM, for using far pointers or accessing
asm symbols. Dj64dev progs switch
to RM infrequently, plus their djgpp
counterparts also switch to RM from
32bit PM, which also gives penalty.
So at least in theory dj64dev progs
should be slightly faster than their
counterparts. But w/o threading this
all doesn't matter: DOS is not about
performance.

Понедельник, 30 декабря 2024 г получено от Andrew Bird:

**Comment by jschwartzenberg** [comment]
Btw. it's great to see dj64dev being so mature. Do you have any programs in mind to build with it? And maybe ship with packages along in the FreeDOS repos?
Two ideas I have:
- Look which programs now installed with install-freedos are built with djgpp
- Generally make a list of DJGPP buildable programs and try building them with dj64dev

Which advantages do you expect to see when building with dj64dev over djgpp when running in dosemu2?

**Reply by stsp** [reply]
> Btw. it's great to see dj64dev being so mature.

3 years of work. :)

> Do you have any programs in mind to build with it?

Gopherus.
Which is why I was picking into tcp/ip
as well. But the good tcp/ip is not yet
there, so Gopherus is waiting. Maybe
someone else wants to do tcp/ip work
in a mean time? :)

> And maybe ship with packages along in the FreeDOS repos?

Nope - they are non-functional w/o
dosemu2. At least until someone
writes a sufficiently capable DOS
extender. Which means never. :)

> Look which programs now installed with install-freedos are built with djgpp

Likely none.
AFAIK freedos avoids djgpp, prefering
the 16bit compilers.

> Generally make a list of DJGPP buildable programs

Its here:
https://www.math.utah.edu/docs/info/gnuish_1.html

> and try building them with dj64dev

Most of them are not interesting for
dosemu2 directly. They belong to freedos.
I find Gopherus an exception of that
rule, and it likely deserves to be packaged
with dosemu2 on its own. There may
be other such packages: some GUI,
maybe some tetris. :) We can have
our own repo of progs, but it should
be small.

**Reply by stsp** [reply]
So why don't you add gnuish into
your installer scripts?
https://www.ibiblio.org/pub/micro/pc-stuff/freedos/mirrors/gnuish/
